stubs/openpyxl/openpyxl/workbook/workbook.pyi

```diff
@@ -20,54 +30,54 @@ class Workbook:
     @property
-    def write_only(self): ...
+    def write_only(self) -> bool: ...
     @property
-    def excel_base_date(self): ...
+    def excel_base_date(self) -> datetime: ...
     @property
-    def active(self): ...
+    def active(self) -> _WorkbookChild | None: ...
     @active.setter
-    def active(self, value) -> None: ...
-    def create_sheet(self, title: Incomplete | None = ..., index: Incomplete | None = ...): ...
-    def move_sheet(self, sheet, offset: int = ...) -> None: ...
-    def remove(self, worksheet) -> None: ...
-    def remove_sheet(self, worksheet) -> None: ...
-    def create_chartsheet(self, title: Incomplete | None = ..., index: Incomplete | None = ...): ...
-    def get_sheet_by_name(self, name): ...
-    def __contains__(self, key): ...
-    def index(self, worksheet): ...
-    def get_index(self, worksheet): ...
-    def __getitem__(self, key): ...
-    def __delitem__(self, key) -> None: ...
-    def __iter__(self): ...
-    def get_sheet_names(self): ...
+    def active(self, value: _WorkbookChild | int) -> None: ...
+    def create_sheet(self, title: str | None = ..., index: int | None = ...): ...
+    def move_sheet(self, sheet: Worksheet | str, offset: int = ...) -> None: ...
+    def remove(self, worksheet: Worksheet) -> None: ...
+    def remove_sheet(self, worksheet: Worksheet) -> None: ...
+    def create_chartsheet(self, title: str | None = ..., index: int | None = ...) -> Chartsheet: ...
+    def get_sheet_by_name(self, name: str) -> Worksheet: ...
+    def __contains__(self, key: str) -> bool: ...
+    def index(self, worksheet: Worksheet) -> int: ...
+    def get_index(self, worksheet: Worksheet) -> int: ...
+    def __getitem__(self, key: str) -> Worksheet: ...
+    def __delitem__(self, key: str) -> None: ...
+    def __iter__(self) -> Iterator[Worksheet]: ...
+    def get_sheet_names(self) -> list[Worksheet]: ...
     @property
-    def worksheets(self): ...
+    def worksheets(self) -> list[Worksheet]: ...
     @property
-    def chartsheets(self): ...
+    def chartsheets(self) -> list[Chartsheet]: ...
     @property
-    def sheetnames(self): ...
+    def sheetnames(self) -> list[str]: ...
     def create_named_range(
-        self, name, worksheet: Incomplete | None = ..., value: Incomplete | None = ..., scope: Incomplete | None = ...
+        self, name: str, worksheet: Worksheet | None = ..., value: str | Incomplete | None = ..., scope: Incomplete | None = ...
     ) -> None: ...
-    def add_named_style(self, style) -> None: ...
+    def add_named_style(self, style: NamedStyle) -> None: ...
     @property
-    def named_styles(self): ...
-    def get_named_ranges(self): ...
-    def add_named_range(self, named_range) -> None: ...
-    def get_named_range(self, name): ...
-    def remove_named_range(self, named_range) -> None: ...
+    def named_styles(self) -> list[str]: ...
+    def get_named_ranges(self) -> list[DefinedName] | tuple[DefinedName]: ...
```

Comment:
```suggestion
    def get_named_ranges(self) -> list[DefinedName] | tuple[DefinedName, ...]: ...
```

Remember that `tuple[T]` means "a single-element tuple where the sole element is of type `T`". You want `tuple[T, ...]` if you want to signify "a tuple of arbitrary length, where all elements are of type `T`".

We could also possibly simplify this by just doing `Sequence[DefinedName]` (with `Sequence` imported from `collections.abc`)?

```suggestion
    def get_named_ranges(self) -> Sequence[DefinedName]: ...
```

Reply:
`collections.abc.Sequence` is broader than just `list` and `tuple`. It also includes `str` and `bytes`: [https://docs.python.org/3/glossary.html#term-sequence](https://docs.python.org/3/glossary.html#term-sequence)

I don't think `collections.abc.Sequence` would be appropriate.

Reply:
> `collections.abc.Sequence` is broader than just `list` and `tuple`. It also includes `str` and `bytes`: https://docs.python.org/3/glossary.html#term-sequence

`str` and `bytes` are indeed both sequences, correct — but neither is a subtype of `Sequence[DefinedName]`. `str` is a subtype of `Sequence[str]`, and `bytes` is a subtype of `Sequence[int]`. So if I were writing this PR, I would favour the simpler approach here :)

But your approach isn't wrong by any means, and I don't want to be too picky :)

Reply:
2 things here I just noticed (sorry it's been a while that I've looked at openpyxl)
1: descriptors are not meant to be used directly for type hints. Descriptors are assigned to ClassVars of classes subtyping Serializable. Serializable members act as properties where the setter uses the descriptor for runtime vlaidation of the setter
2. Whilst the `Sequence` descriptor allows both list and tuple (https://foss.heptapod.net/openpyxl/openpyxl/-/blob/branch/3.1/openpyxl/descriptors/sequence.py#L18), `DefinedNameList.definedName` specifically needs to be sliceable and appendable (at least in 3.0, this doesn't seem to be the case anymore for 3.1) https://foss.heptapod.net/openpyxl/openpyxl/-/blob/branch/3.0/openpyxl/workbook/defined_name.py#L207-209, the latter meaning it can't be a tuple.